ASoC: SOF: Intel: Fix kernel crash caused by surprisingly broken firmware

ujfalusi opened the pull request:
Hi,

@kv2019i managed to 'craft' a firmware which for some unknown reasons sends a reply message before the FW_READY (it might not be the firmware, but the firmware causes this).
Since the reply_data is allocated after the first fw_ready message we will hit a NULL pointer in kernel, which is not a nice thing to do, even if the user loads broken binary to the DSP.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -61,17 +61,23 @@ irqreturn_t cnl_ipc4_irq_thread(int irq, void *context)
 			/* Reply received */
-			struct sof_ipc4_msg *data = sdev->ipc->msg.reply_data;
+			if (likely(sdev->fw_state == SOF_FW_BOOT_COMPLETE)) {
+				struct sof_ipc4_msg *data = sdev->ipc->msg.reply_data;
```
Commit message typo in all  commits

the same condition ~~can~~ is present for IPC3.

ujfalusi commented:
Changes since v1:
- commit messages updated

ranj063 commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -61,17 +61,23 @@ irqreturn_t cnl_ipc4_irq_thread(int irq, void *context)
 
 		if (primary & SOF_IPC4_MSG_DIR_MASK) {
 			/* Reply received */
-			struct sof_ipc4_msg *data = sdev->ipc->msg.reply_data;
```
@ujfalusi just out of curiosity, why do we need the warning if we're going to ignore this anyway? I feel like this will come back to bite us later on

ujfalusi replied:
@ranj063, we use it to print the warning as we need to know about this but not going to process it as a reply to avoid NULL pointer
We had similar issue with IMR boot, but since that bitten only after the first boot, we had the reply_data allocated already.

ranj063 replied:
@ujfalusi if you have this warning and it so happens that we see it during first boot, what do we plan to do with it?

ujfalusi replied:
@ranj063, well, if we know that we have received `0x70000000|0x80` before fw_ready and then we have the dsp dump with ROM states (which @kv2019i omitted from the log), we might be able to figure out what is going on?

ranj063 replied:
@ujfalusi i think we should downgrade this to a debug instead of warning. Anyway, if others are OK with a warning, you still have a typo in all your commit messages duplicated a "a a"

kv2019i commented:
Seems to be working. With the faulty FW, I now get this and no kernel panic anymore:

UPDATE @ujfalusi  more complete log
```
[   42.225445] sof-audio-pci-intel-tgl 0000:00:1f.3: Attempting iteration 0 of Core En/ROM load...
[   42.225451] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x3010303 successful
[   42.225461] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x3010302 successful
[   42.225462] sof-audio-pci-intel-tgl 0000:00:1f.3: unstall/run core: core_mask = 1
[   42.225465] sof-audio-pci-intel-tgl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask 1
[   42.225975] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0xd4]=0x80000000 successful
[   42.279395] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5000001 successful
[   42.297900] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5 successful
[   42.297904] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x140000 successful
[   42.297906] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware download successful, booting...
[   42.309633] sof-audio-pci-intel-tgl 0000:00:1f.3: IPC reply before FW_READY: 0x70000000|0x80
[   44.360377] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump start ]------------
[   44.360382] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware boot failure due to timeout
[   44.360383] sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state: SOF_FW_BOOT_IN_PROGRESS (2)
[   44.360392] sof-audio-pci-intel-tgl 0000:00:1f.3: status: fw entered - code 00000005
[   44.360482] sof-audio-pci-intel-tgl 0000:00:1f.3: invalid header size 0x3010202. FW oops is bogus
[   44.360483] sof-audio-pci-intel-tgl 0000:00:1f.3: unexpected fault 0x00000000 trace 0x00004000
[   44.360485] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump end ]------------
[   44.360487] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to boot DSP firmware -5
[   44.360488] sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 2 -> 3
[   44.362240] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x3010303 successful
[   44.362776] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x303 successful
[   44.362781] sof-audio-pci-intel-tgl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask 1
[   44.363119] sof-audio-pci-intel-tgl 0000:00:1f.3: display power disable
[   44.364098] sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 3 -> 0
```

kv2019i reviewed (approved):
Thanks @ujfalusi ! I tried to attach the fault-inducing-fw, but github refused this, but drop me note if you want to try. I do think this must be fixed as the FW that causes this has an invalid signature (but not alone a sufficient condition to trigger the problem).

RanderWang reviewed (dismissed):
ROM could send ipc message to host. But in the log 0x70000000 | 0x80 is what ?

ujfalusi commented:
> ROM could send ipc message to host. But in the log 0x70000000 | 0x80 is what ?

@RanderWang, that is what puzzles me as well. if I decode it, it will be like
module message (bit 30)
it is reply (bit 29)
the message type is 0x10 which is not even defined.

Unless this is not an IPC message at all, but something else or decoded completely differently...

ujfalusi commented:
But according the the FSR code. the firmware is entered, so it must be something that the firmware is sending.
CSE decrypted the firmware even with a wrong key and we got something which apparently run or something and does not too sane thing?

ujfalusi commented:
Changes since v2:
- typo fixed in commit messages (double a)
- downgraded the prints to dev_dbg_ratelimited() since this issue should not happen on user's machine, only during development.
